File: mobile/native/android/apps/ensu/data/src/main/java/io/ente/ensu/data/llm/EnsuRustDefaults.kt
package io.ente.ensu.data.llm

import io.ente.ensu.domain.model.EnsuDefaults
import io.ente.ensu.domain.model.EnsuModelPreset
import io.ente.ensu.bindings.EnsuModelPreset as NativeEnsuModelPreset
import io.ente.ensu.bindings.getEnsuDefaults
import io.ente.ensu.bindings.uniffiEnsureInitialized

object EnsuRustDefaults {
    fun load(): EnsuDefaults {
        uniffiEnsureInitialized()
        val defaults = getEnsuDefaults()
        return EnsuDefaults(
            mobileSystemPromptBody = defaults.mobileSystemPromptBody,
            desktopSystemPromptBody = defaults.desktopSystemPromptBody,
            systemPromptDatePlaceholder = defaults.systemPromptDatePlaceholder,
            sessionSummarySystemPrompt = defaults.sessionSummarySystemPrompt,
            mobileDefaultModel = defaults.mobileDefaultModel.toDomain(),
            mobileModelPresets = defaults.mobileModelPresets.map { it.toDomain() },
            desktopDefaultModel = defaults.desktopDefaultModel.toDomain(),
            desktopModelPresets = defaults.desktopModelPresets.map { it.toDomain() }
        )
    }
}

private fun NativeEnsuModelPreset.toDomain(): EnsuModelPreset =
    EnsuModelPreset(
        id = id,
        title = title,
        url = url,
        mmprojUrl = mmprojUrl
    )

File: mobile/native/darwin/Apps/Ensu/Ensu/Services/EnsuRustDefaults.swift
import Foundation

struct EnsuRustModelPreset: Equatable {
    let id: String
    let title: String
    let url: String
    let mmprojUrl: String?
}

struct EnsuRustDefaultsValue {
    let mobileSystemPromptBody: String
    let desktopSystemPromptBody: String
    let systemPromptDatePlaceholder: String
    let sessionSummarySystemPrompt: String
    let mobileDefaultModel: EnsuRustModelPreset
    let mobileModelPresets: [EnsuRustModelPreset]
    let desktopDefaultModel: EnsuRustModelPreset
    let desktopModelPresets: [EnsuRustModelPreset]
}

enum EnsuRustDefaults {
    static let shared: EnsuRustDefaultsValue = load()

    private static func load() -> EnsuRustDefaultsValue {
        let defaults = getEnsuDefaults()
        return EnsuRustDefaultsValue(
            mobileSystemPromptBody: defaults.mobileSystemPromptBody,
            desktopSystemPromptBody: defaults.desktopSystemPromptBody,
            systemPromptDatePlaceholder: defaults.systemPromptDatePlaceholder,
            sessionSummarySystemPrompt: defaults.sessionSummarySystemPrompt,
            mobileDefaultModel: defaults.mobileDefaultModel.toLocal(),
            mobileModelPresets: defaults.mobileModelPresets.map { $0.toLocal() },
            desktopDefaultModel: defaults.desktopDefaultModel.toLocal(),
            desktopModelPresets: defaults.desktopModelPresets.map { $0.toLocal() }
        )
    }
}

private extension EnsuModelPreset {
    func toLocal() -> EnsuRustModelPreset {
        EnsuRustModelPreset(
            id: id,
            title: title,
            url: url,
            mmprojUrl: mmprojUrl
        )
    }
}
